stackrox/collector
# Use inotify for reloading configuration in all cases

## Description

Stop polling for the runtime configuration file and instead fully rely on inotify for reloading. In order to achieve this, 2 new classes are introduced:

- Inotify: handles operations involving inotify, including:
  - Opening and closing the inotify file descriptor.
  - Adding, removing and keeping track of file/directory watchers.
  - Reading events and forwarding them to a caller in a non-blocking fashion.
- ConfigLoader: uses an Inotify object for tracking changes to the runtime configuration file and update it accordingly.

## Checklist

**Automated testing**

If any of these don't apply, please comment below.

## Testing Performed

Tested manually on a kind cluster.

Two k8s integration tests were added using the configuration introspection endpoint:
- First test:
  - Launch collector with a config map already created.
  - Check the external IPs feature has the correct setting.
  - Edit the configmap and check collector reloads the configuration correctly.
- Second test:
  - Launch collector with no configmap and check runtime configuration is empty.
  - Create the configuration map and check collector starts using runtime configuration with the correct value.
  - Edit the configmap and check collector reloads the configuration correctly.
  - Remove the configmap and check collector drops runtime configuration.
